### PR TITLE
Support kubernetes group api

### LINKED ISSUE #144: Support new group API
Kubernetes documentation:

- http://kubernetes.io/v1.1/docs/proposals/api-group.html#server-side-implementation

following the kubernetes description the entities can be reached using the `prefix`, `group` and `version` values.

- for `/api/v1`, we set the `prefix="api"`, `group=""`, `version="v1"`, so the URL remains to be `/api/v1`

A similar approach (client-side) could be used in kubeclient.

- `prefix` and `group` should be preconfigured per-entity (e.g. for all current entities should be `prefix="api"`, `group=""`)
- `version` should be more easily configurable per request

For example:

    client.get_pods
    # prefix="api", group="", version="v1"
    # => /api/v1/pods

For group api (e.g. deployments) we could use just a plain form in the API:

- `client.get_deployments`

but I believe that it would be better to scope them in some way:

- `client.extensions.get_deployments`
- `client.extensions_get_deployments`

and for example:

    client.extensions.get_deployments
    # prefix="apis", group="extensions", version="v1beta1"
    # => /apis/extensions/v1beta1/depoyments

### PR DESCRIPTION
**Pull request by moolitayer**
[kubernetes api-group proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/api-group.md)
[kubernetes api](https://github.com/kubernetes/kubernetes/blob/master/docs/api.md)

[specification](https://github.com/abonas/kubeclient/issues/144#issuecomment-235254248)

[Some testing code](https://gist.github.com/moolitayer/476ed01d377fc9e9954fea83572ba662)
(Devided into core, groups & OpenShift)

Fixes #144 
Does not handle nested (e.g pods/status) and first level (e.g /healthz) resources

### CODE REVIEW

**Comment by simon3z** [comment]
@enoodle @zeari @yaacov @zakiva please review.

File: lib/kubeclient/common.rb [hunk 1]
```diff
@@ -71,68 +80,52 @@ def handle_exception
...
-          update_entity(entity_type, entity_config)
-        end
-
-        define_method("patch_#{entity_name}") do |name, patch, namespace = nil|
-          patch_entity(entity_type, name, patch, namespace)
+          args = [singular, options]
+        else
+          args = [singular] + args
         end
+        puts "calling: #{method_name}(#{args})"
```
Comment (enoodle) [inline lib/kubeclient/common.rb]: Are we sure that we want to have this ```puts``` here?

File: lib/kubeclient/common.rb [hunk 2]
```diff
@@ -267,8 +261,23 @@ def patch_entity(entity_type, name, patch, namespace = nil)
       end
     end
 
-    def new_entity(hash, klass)
-      klass.new(hash)
+    def new_entity(hash, entity_type)
+      resource_class(entity_type).new(hash)
+    end
+
+    def resource_class(entity_type)
```
Comment (enoodle) [inline lib/kubeclient/common.rb]: :+1:

File: lib/kubeclient/common.rb [hunk 3]
```diff
@@ -267,8 +261,23 @@ def patch_entity(entity_type, name, patch, namespace = nil)
...
+      # TODO: test binding since there already seems to be a constant with that name
+      Kubeclient.const_get(entity_type)
+      rescue NameError
+        Kubeclient.const_set(
```
Comment (enoodle) [inline lib/kubeclient/common.rb]: Can we check if the entity actually exists?

**Comment by enoodle** [comment]
As far as I understand this doesn't break the old API at all. e.g the class definitions can still be used. Am I right?

**Comment by moolitayer** [comment]
> As far as I understand this doesn't break the old API at all. e.g the class definitions can still be used. Am I right?

I'm afraid It does, I've exapalained why it is problematic to maintain that backward compitability in this [comment](https://github.com/abonas/kubeclient/issues/144#issuecomment-235254248) see Discovery

File: lib/kubeclient/common.rb (hunk 3 above)
Comment (moolitayer) [inline lib/kubeclient/common.rb]: > Can we check if the entity actually exists?

When a request is made for a resource that does not exist we raise a [name error](https://github.com/abonas/kubeclient/pull/181/files#diff-104b3d1973988933ffc83feada470da2R71).
That is done per your previous suggestion, to be backward compatible.

**Comment by moolitayer** [comment]
@simon3z @abonas please review

File: lib/kubeclient/common.rb [hunk 4]
```diff
@@ -68,6 +68,7 @@ def handle_exception
         {}
       end
       err_message = json_error_msg['message'] || e.message
+      raise NameError.new if json_error_msg && json_error_msg['reason'] == 'NotFound'
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @moolitayer isn't "NotFound" used also when you look for a specific entity and it's not found? (e.g. `get_pod("my_pod")`).

File: lib/kubeclient/common.rb [hunk 5]
```diff
@@ -76,57 +77,49 @@ def handle_uri(uri, path)
...
+        options = { resource_version: options } unless options.is_a?(Hash)
+        args = [singular, options]
+      else
+        args = [singular] + args
+      end
+      send(method_name.to_sym, *args)
+    end
+
+    def respond_to?(method_sym)
+      resource_method?(method_sym) ? true : super
```
Comment (abonas) [inline lib/kubeclient/common.rb]: who would be the `super` here?

File: lib/kubeclient/common.rb (hunk 4 above)
Comment (moolitayer) [inline lib/kubeclient/common.rb]: Right :(
The error message is different. I'm wondering If we could use that?
For non existing types we get:
```"message": "the server could not find the requested resource",```
For non existing entities:
```"message": "nodes \"127.0.0.12\" not found",```

File: lib/kubeclient/common.rb (hunk 5 above)
Comment (moolitayer) [inline lib/kubeclient/common.rb]: The including object;
```
> client.hi
NoMethodError: undefined method `hi' for #<Kubeclient::Client:0x00000003ed5118>
```

**Comment by simon3z** [comment]
@moolitayer can you rebase this? I want to merge this ASAP so that it will be tested by everyone longer.

**Comment by moolitayer** [comment]
This is more then a rebase. The approach would need to change since it relay's on activesupport (with a robust implementation to handle entities we now know nothing about). I think there are solutions and I will be looking into this.

**Comment by moolitayer** [comment]
Now that #184 was merged I'm dropping this solution in favor of #185

**Pull request closed without merging**